Force elements in MultibodyTree

amcastro-tri opened the pull request:
This PR introduces the `ForceElement` class which allows to define general force models that depend on state such as spring, dampers, drag, and gravity.
Unit tests in `double_pendulum_test.cc`.

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/robotlocomotion/drake/6957)
<!-- Reviewable:end -->

amcastro-tri commented:
Reviewed 14 of 26 files at r1, 5 of 12 files at r2.
Review status: all files reviewed at latest revision, all discussions resolved, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/6957)*
<!-- Sent from Reviewable.io -->

amcastro-tri commented:
Reviewed 1 of 4 files at r3.
Review status: 17 of 20 files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/6957)*
<!-- Sent from Reviewable.io -->

amcastro-tri commented:
This is now dependent on #6958. Will come back to this after #6958 gets merged.

---

Review status: 16 of 20 files reviewed at latest revision, all discussions resolved, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/6957#-:-KszrdNNQ7ePyD079zMp:bt03l5w)*
<!-- Sent from Reviewable.io -->

amcastro-tri commented:
Reviewed 2 of 4 files at r3, 12 of 13 files at r4.
Review status: all files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/6957)*
<!-- Sent from Reviewable.io -->

amcastro-tri commented:
Reviewed 5 of 5 files at r5.
Review status: all files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/6957)*
<!-- Sent from Reviewable.io -->

amcastro-tri commented:
+@edrumwri for feature review please.

---

Review status: all files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/6957#-:-KuV1KuFc3FslDIVKhbZ:b-a98c4g)*
<!-- Sent from Reviewable.io -->

jwnimmer-tri commented:
@amcastro-tri when you are the author, you should _not_ mark the green eyeball in reviewable, or else it will confuse the discussion workflow.  You should go through and unset the eyeball on every file.

---







*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/6957#-:-KuV1z8zpzmtpSnNe_GP:bk4w1d5)*
<!-- Sent from Reviewable.io -->

amcastro-tri commented:
Thanks @jwnimmer-tri. I really didn't know it could cause any confusion. This was just me double-checking changes after a big merge. I unchecked the files now.

---

Review status: 0 of 17 files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/6957#-:-KuV37aaBQEtw3rEDTru:b-rahgan)*
<!-- Sent from Reviewable.io -->

edrumwri commented:
First pass complete but for uniform_gravity_field_element.cc

---

Reviewed 1 of 26 files at r1, 1 of 12 files at r2, 3 of 4 files at r3, 7 of 13 files at r4, 4 of 5 files at r5.
Review status: 16 of 17 files reviewed at latest revision, 17 unresolved discussions, some commit checks failed.

---

*[drake/multibody/benchmarks/acrobot/acrobot.cc, line 96 at r5](https://reviewable.io:443/reviews/robotlocomotion/drake/6957#-Ku_MN-FLankStt4jesD:-Ku_MN-GaJ0Fq6Y9nxcM:b-z8q6gt) ([raw file](https://github.com/robotlocomotion/drake/blob/1faa309e4a3f83aaba3ea1508a6d68f273fa76a8/drake/multibody/benchmarks/acrobot/acrobot.cc#L96)):*
> ```cpp
>   // Unlike http://underactuated.mit.edu/underactuated.html?chapter=3, we define
>   // G(q) to be on the right hand side of the equations of motion, that is,
>   // Mv̇ + C(q, v)v = G(q).
> ```

Alejandro, we have a name clash here- g() is often used for constraint functions and G(q), in particular, is frequently used for the Jacobian matrices of these constraint functions. What do you think about replacing C(q, v)v with c(q, v) to incorporate all non-mass related terms? 

---

*[drake/multibody/benchmarks/acrobot/acrobot.h, line 80 at r5](https://reviewable.io:443/reviews/robotlocomotion/drake/6957#-Ku_MsmBxxDowamOcVCo:-Ku_MsmBxxDowamOcVCp:b15ygub) ([raw file](https://github.com/robotlocomotion/drake/blob/1faa309e4a3f83aaba3ea1508a6d68f273fa76a8/drake/multibody/benchmarks/acrobot/acrobot.h#L80)):*
> ```objc
>   /// Unlike http://underactuated.mit.edu/underactuated.html?chapter=3, cited in
>   /// this class's documentation, we define `G(q)` to be on the right hand side
>   /// of the equations of motion, that is, `Mv̇ + C(q, v)v = G(q)`.
> ```

Same name collision with G(q) here.

---

*[drake/multibody/multibody_tree/body.h, line 172 at r5](https://reviewable.io:443/reviews/robotlocomotion/drake/6957#-Ku_N3ViVavLwdn8fgNt:-Ku_N3ViVavLwdn8fgNu:b-nujelk) ([raw file](https://github.com/robotlocomotion/drake/blob/1faa309e4a3f83aaba3ea1508a6d68f273fa76a8/drake/multibody/multibody_tree/body.h#L172)):*
> ```objc
> 
>   /// Computes the mass of this body at the given state described by `context`.
>   virtual T CalcMass(const MultibodyTreeContext<T>& context) const = 0;
> ```

Why might the mass change?

---

*[drake/multibody/multibody_tree/force_element.h, line 18 at r5](https://reviewable.io:443/reviews/robotlocomotion/drake/6957#-KuaeRxsfO0FdD71Lqcw:-KuaeRxsfO0FdD71Lqcx:bk0xw7a) ([raw file](https://github.com/robotlocomotion/drake/blob/1faa309e4a3f83aaba3ea1508a6d68f273fa76a8/drake/multibody/multibody_tree/force_element.h#L18)):*
> ```objc
> namespace multibody {
> 
> /// A %ForceElement allows to model state dependent forces in a MultibodyTree
> ```

BTW allows to model -> allows modeling

---

*[drake/multibody/multibody_tree/force_element.h, line 19 at r5](https://reviewable.io:443/reviews/robotlocomotion/drake/6957#-KuaeZa0mXvVLDhhs6Ug:-KuaeZa0mXvVLDhhs6Uh:b-hiwvtf) ([raw file](https://github.com/robotlocomotion/drake/blob/1faa309e4a3f83aaba3ea1508a6d68f273fa76a8/drake/multibody/multibody_tree/force_element.h#L19)):*
> ```objc
> 
> /// A %ForceElement allows to model state dependent forces in a MultibodyTree
> /// model. Examples of such forces are springs, dampers, drag and gravity.
> ```

What about time dependent forces?

---

*[drake/multibody/multibody_tree/force_element.h, line 44 at r5](https://reviewable.io:443/reviews/robotlocomotion/drake/6957#-Kuaf82bSgD8RSW-HegM:-Kuaf82bSgD8RSW-HegN:bm6xbyw) ([raw file](https://github.com/robotlocomotion/drake/blob/1faa309e4a3f83aaba3ea1508a6d68f273fa76a8/drake/multibody/multibody_tree/force_element.h#L44)):*
> ```objc
> 
>   /// Computes the force contribution for `this` force element and **adds** it
>   /// to the output arrays of forces. Depending on their model, different force
> ```

BTW Cool!

---

*[drake/multibody/multibody_tree/force_element.h, line 68 at r5](https://reviewable.io:443/reviews/robotlocomotion/drake/6957#-KuafUHhNc7u9Y_6Lrux:-KuafUHhNc7u9Y_6Lruy:b-qa4h9r) ([raw file](https://github.com/robotlocomotion/drake/blob/1faa309e4a3f83aaba3ea1508a6d68f273fa76a8/drake/multibody/multibody_tree/force_element.h#L68)):*
> ```objc
>   /// @param[out] tau_array
>   ///   On output `this` force element adds its contribution to the total
>   ///   generalized forces. Typically used to model generalized forces that
> ```

BTW This sentence is repetitive- perhaps change "to model generalized forces" to "to model force contributions"

---

*[drake/multibody/multibody_tree/force_element.h, line 70 at r5](https://reviewable.io:443/reviews/robotlocomotion/drake/6957#-Kuafd1zoiE4gOyTVqOg:-Kuafd2-k66K9ym4aTRo:b-7c5u9r) ([raw file](https://github.com/robotlocomotion/drake/blob/1faa309e4a3f83aaba3ea1508a6d68f273fa76a8/drake/multibody/multibody_tree/force_element.h#L70)):*
> ```objc
>   ///   generalized forces. Typically used to model generalized forces that
>   ///   are best formulated as generalized forces rather than as spatial forces.
>   ///   For instance, damping on a RevoluteMobilizer's angular velocity.
> ```

BTW Sentence fragment.

---

*[drake/multibody/multibody_tree/force_element.h, line 137 at r5](https://reviewable.io:443/reviews/robotlocomotion/drake/6957#-KuagfI8bapvAZBVD_wO:-KuagfI8bapvAZBVD_wP:b-y6ojat) ([raw file](https://github.com/robotlocomotion/drake/blob/1faa309e4a3f83aaba3ea1508a6d68f273fa76a8/drake/multibody/multibody_tree/force_element.h#L137)):*
> ```objc
>   /// defined to be positive when the potential energy is decreasing. In other
>   /// words, if `PE` is the potential energy as defined by
>   /// CalcPotentialEnergy(), then the conservative power is `Pc = -d(PE)/dt`.
> ```

FYI Nit: "then the conservative power, Pc, is"

---

*[drake/multibody/multibody_tree/force_element.h, line 140 at r5](https://reviewable.io:443/reviews/robotlocomotion/drake/6957#-KuahkkI64hV7ZqLnPzo:-KuahkkI64hV7ZqLnPzp:bl03x7k) ([raw file](https://github.com/robotlocomotion/drake/blob/1faa309e4a3f83aaba3ea1508a6d68f273fa76a8/drake/multibody/multibody_tree/force_element.h#L140)):*
> ```objc
>   ///
>   /// @see CalcPotentialEnergy(), CalcNonConservativePower()
>   virtual T CalcConservativePower(
> ```

If nothing else, the lack of symmetry in having the user define NVI and non-NVI methods (i.e., DoX vs. X) is a little jarring.

---

*[drake/multibody/multibody_tree/force_element.h, line 173 at r5](https://reviewable.io:443/reviews/robotlocomotion/drake/6957#-KuagzDWdLb9hdKMSp7E:-KuagzDXCeuH0dA5HSdq:b-88o0wo) ([raw file](https://github.com/robotlocomotion/drake/blob/1faa309e4a3f83aaba3ea1508a6d68f273fa76a8/drake/multibody/multibody_tree/force_element.h#L173)):*
> ```objc
> 
>  protected:
>   virtual void DoCalcAndAddForceContribution(
> ```

Please ref CalcAndAddForceContribution() for a description for the parameters of this function.

---

*[drake/multibody/multibody_tree/force_element.h, line 185 at r5](https://reviewable.io:443/reviews/robotlocomotion/drake/6957#-KuahRnklbXAWc-0uSKv:-KuahRnklbXAWc-0uSKw:boaylju) ([raw file](https://github.com/robotlocomotion/drake/blob/1faa309e4a3f83aaba3ea1508a6d68f273fa76a8/drake/multibody/multibody_tree/force_element.h#L185)):*
> ```objc
>   /// Clones this %ForceElement (templated on T) to a mobilizer templated on
>   /// `double`.
>   virtual std::unique_ptr<ForceElement<double>> DoCloneToScalar(
> ```

In your great description at the beginning of this class, you do not describe these two functions as needing to be implemented.

---

*[drake/multibody/multibody_tree/multibody_tree.h, line 400 at r5](https://reviewable.io:443/reviews/robotlocomotion/drake/6957#-KuaifiqKbl0M8kK8vvl:-KuaifiqKbl0M8kK8vvm:bmcuzic) ([raw file](https://github.com/robotlocomotion/drake/blob/1faa309e4a3f83aaba3ea1508a6d68f273fa76a8/drake/multibody/multibody_tree/multibody_tree.h#L400)):*
> ```objc
>   }
> 
>   template<template<typename Scalar> class ForceElementType, typename... Args>
> ```

BTW Nice!

---

*[drake/multibody/multibody_tree/multibody_tree.h, line 781 at r5](https://reviewable.io:443/reviews/robotlocomotion/drake/6957#-Kuail_25CbivuPwXyCo:-Kuail_3DnZmFCEYxQET:bqqjxz7) ([raw file](https://github.com/robotlocomotion/drake/blob/1faa309e4a3f83aaba3ea1508a6d68f273fa76a8/drake/multibody/multibody_tree/multibody_tree.h#L781)):*
> ```objc
>   /// method provides outputs for both spatial forces per body (with
>   /// `F_Bo_W_array`) and generalized forces (with `tau_array`).
>   /// ForceElement contributions are a function of the state only.
> ```

Again, why can't they be a function of time?

---

*[drake/multibody/multibody_tree/multibody_tree.h, line 1033 at r5](https://reviewable.io:443/reviews/robotlocomotion/drake/6957#-KuajdF85NnNrxB3bln7:-KuajdF85NnNrxB3bln8:b-90t96q) ([raw file](https://github.com/robotlocomotion/drake/blob/1faa309e4a3f83aaba3ea1508a6d68f273fa76a8/drake/multibody/multibody_tree/multibody_tree.h#L1033)):*
> ```objc
>   // tree.
>   template <typename FromScalar>
>   ForceElement<T>* CloneForceElementAndAdd(
> ```

Any reason you don't return a const reference as you did with AddForceElement() above? Actually, given that this is a helper method, why return anything at all?

---

*[drake/multibody/multibody_tree/uniform_gravity_field_element.h, line 13 at r5](https://reviewable.io:443/reviews/robotlocomotion/drake/6957#-KuakOQ-lmOq_kQz46b6:-KuakOQ0Ft8V6ynzjTlC:b-1va35w) ([raw file](https://github.com/robotlocomotion/drake/blob/1faa309e4a3f83aaba3ea1508a6d68f273fa76a8/drake/multibody/multibody_tree/uniform_gravity_field_element.h#L13)):*
> ```objc
> namespace multibody {
> 
> /// This ForceElement allows to model the effect of a uniform gravity field as
> ```

BTW "allows to model" -> "allows modeling"

---

*[drake/multibody/multibody_tree/uniform_gravity_field_element.h, line 36 at r5](https://reviewable.io:443/reviews/robotlocomotion/drake/6957#-KuakZnqtO9q4BPrwM8M:-KuakZnqtO9q4BPrwM8N:b60xoa) ([raw file](https://github.com/robotlocomotion/drake/blob/1faa309e4a3f83aaba3ea1508a6d68f273fa76a8/drake/multibody/multibody_tree/uniform_gravity_field_element.h#L36)):*
> ```objc
>   /// Returns the acceleration of gravity vector, expressed in the world frame
>   /// W.
>   const Vector3<double>& g_W() const { return g_W_; }
> ```

I'll posit that `g_W` violates the style guide on naming conventions.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/6957#-:-KuanG1sH7wc3XWIDncZ:bimcdiu)*
<!-- Sent from Reviewable.io -->

edrumwri commented:
Reviewed 1 of 5 files at r5.
Review status: all files reviewed at latest revision, 18 unresolved discussions, some commit checks failed.

---

*[drake/multibody/multibody_tree/uniform_gravity_field_element.cc, line 103 at r5](https://reviewable.io:443/reviews/robotlocomotion/drake/6957#-KuawCBjKXEzSF-gtaLJ:-KuawCBl743p90fFGdvv:b-kihi3m) ([raw file](https://github.com/robotlocomotion/drake/blob/1faa309e4a3f83aaba3ea1508a6d68f273fa76a8/drake/multibody/multibody_tree/uniform_gravity_field_element.cc#L103)):*
> ```cpp
>     // The conservative power is defined to be positive when the potential
>     // energy decreases.
>     TotalConservativePower += (mass * v_WBcm.dot(g_W()));
> ```

Please add a test of this function.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/6957)*
<!-- Sent from Reviewable.io -->

sherm1 commented:
Review status: all files reviewed at latest revision, 18 unresolved discussions, some commit checks failed.

---

*[drake/multibody/benchmarks/acrobot/acrobot.cc, line 96 at r5](https://reviewable.io:443/reviews/robotlocomotion/drake/6957#-Ku_MN-FLankStt4jesD:-Kub-fBPCFanFcPp_Xrc:b-q3behw) ([raw file](https://github.com/robotlocomotion/drake/blob/1faa309e4a3f83aaba3ea1508a6d68f273fa76a8/drake/multibody/benchmarks/acrobot/acrobot.cc#L96)):*
<details><summary><i>Previously, edrumwri (Evan Drumwright) wrote…</i></summary><blockquote>

Alejandro, we have a name clash here- g() is often used for constraint functions and G(q), in particular, is frequently used for the Jacobian matrices of these constraint functions. What do you think about replacing C(q, v)v with c(q, v) to incorporate all non-mass related terms? 
</blockquote></details>

BTW "standard manipulator form" in Russ's book is
`H(q)q̈+C(q,q̇)q̇+G(q)=Bu`
where G is gravity, mapped into generalized forces. Looks like the G here is the same but for the sign and the comment is intended to make sure someone familiar with that doesn't misunderstand the sign. Since we already changed H to M, maybe it would be better to rename G? Our standard notation is to use f for generalized forces and F for spatial forces, with this sign convention:
`Mv̇ + Cv = f+JᵀF`. Not sure how useful it is to keep the Coriolis forces separate, but the robot folks seem to like it that way.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/6957)*
<!-- Sent from Reviewable.io -->

amcastro-tri commented:
Review status: 8 of 17 files reviewed at latest revision, 12 unresolved discussions.

---

*[drake/multibody/benchmarks/acrobot/acrobot.cc, line 96 at r5](https://reviewable.io:443/reviews/robotlocomotion/drake/6957#-Ku_MN-FLankStt4jesD:-Kuu4bLpuf-8cC8ajN-m:b-l01sut) ([raw file](https://github.com/robotlocomotion/drake/blob/1faa309e4a3f83aaba3ea1508a6d68f273fa76a8/drake/multibody/benchmarks/acrobot/acrobot.cc#L96)):*
<details><summary><i>Previously, sherm1 (Michael Sherman) wrote…</i></summary><blockquote>

BTW "standard manipulator form" in Russ's book is
`H(q)q̈+C(q,q̇)q̇+G(q)=Bu`
where G is gravity, mapped into generalized forces. Looks like the G here is the same but for the sign and the comment is intended to make sure someone familiar with that doesn't misunderstand the sign. Since we already changed H to M, maybe it would be better to rename G? Our standard notation is to use f for generalized forces and F for spatial forces, with this sign convention:
`Mv̇ + Cv = f+JᵀF`. Not sure how useful it is to keep the Coriolis forces separate, but the robot folks seem to like it that way.
</blockquote></details>

People seem to also like lower case `g`, though in texts it usually is a bold g. I think `g(q)` is ok. What do you think?

---

*[drake/multibody/multibody_tree/body.h, line 172 at r5](https://reviewable.io:443/reviews/robotlocomotion/drake/6957#-Ku_N3ViVavLwdn8fgNt:-Kuu4pk5qiApXxnCT5Qk:b-h8b9xf) ([raw file](https://github.com/robotlocomotion/drake/blob/1faa309e4a3f83aaba3ea1508a6d68f273fa76a8/drake/multibody/multibody_tree/body.h#L172)):*
<details><summary><i>Previously, edrumwri (Evan Drumwright) wrote…</i></summary><blockquote>

Why might the mass change?
</blockquote></details>

It might be a parameter of your system say, for sys ID. So far only the API is implemented. 

---

*[drake/multibody/multibody_tree/force_element.h, line 19 at r5](https://reviewable.io:443/reviews/robotlocomotion/drake/6957#-KuaeZa0mXvVLDhhs6Ug:-Kuu5OHNnRs6vn-NhPF2:b7kz8ol) ([raw file](https://github.com/robotlocomotion/drake/blob/1faa309e4a3f83aaba3ea1508a6d68f273fa76a8/drake/multibody/multibody_tree/force_element.h#L19)):*
<details><summary><i>Previously, edrumwri (Evan Drumwright) wrote…</i></summary><blockquote>

What about time dependent forces?
</blockquote></details>

That could certainly be modeled with a `ForceElement` (time is in the context). I updated the doc.

---

*[drake/multibody/multibody_tree/force_element.h, line 140 at r5](https://reviewable.io:443/reviews/robotlocomotion/drake/6957#-KuahkkI64hV7ZqLnPzo:-Kuu7KeKzmsWy9ekn5d6:bc0rg6) ([raw file](https://github.com/robotlocomotion/drake/blob/1faa309e4a3f83aaba3ea1508a6d68f273fa76a8/drake/multibody/multibody_tree/force_element.h#L140)):*
<details><summary><i>Previously, edrumwri (Evan Drumwright) wrote…</i></summary><blockquote>

If nothing else, the lack of symmetry in having the user define NVI and non-NVI methods (i.e., DoX vs. X) is a little jarring.
</blockquote></details>

Actually these should be named `DoCalcXXX()` and be private. The `CalcXXX()` would only thake the `context` as argument. However we still do not have caching. What do you suggest?

---

*[drake/multibody/multibody_tree/force_element.h, line 173 at r5](https://reviewable.io:443/reviews/robotlocomotion/drake/6957#-KuagzDWdLb9hdKMSp7E:-Kuu8Pi4gPM3b66A4wlo:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/1faa309e4a3f83aaba3ea1508a6d68f273fa76a8/drake/multibody/multibody_tree/force_element.h#L173)):*
<details><summary><i>Previously, edrumwri (Evan Drumwright) wrote…</i></summary><blockquote>

Please ref CalcAndAddForceContribution() for a description for the parameters of this function.
</blockquote></details>

Done.

---

*[drake/multibody/multibody_tree/force_element.h, line 185 at r5](https://reviewable.io:443/reviews/robotlocomotion/drake/6957#-KuahRnklbXAWc-0uSKv:-Kuu8iGMAprX3qnsdefp:bbzqahu) ([raw file](https://github.com/robotlocomotion/drake/blob/1faa309e4a3f83aaba3ea1508a6d68f273fa76a8/drake/multibody/multibody_tree/force_element.h#L185)):*
<details><summary><i>Previously, edrumwri (Evan Drumwright) wrote…</i></summary><blockquote>

In your great description at the beginning of this class, you do not describe these two functions as needing to be implemented.
</blockquote></details>

I'd prefer not to do so. The `DoCloneToScalar() mechanism I have right now will hopefully get replaced by something similar to what we have in systems, which will also hopefully allow us to template on novel scalar types without having to change update every  source file. Therefore, I'd prefer not leaking this into the main class's API.

---

*[drake/multibody/multibody_tree/multibody_tree.h, line 400 at r5](https://reviewable.io:443/reviews/robotlocomotion/drake/6957#-KuaifiqKbl0M8kK8vvl:-Kuu9KcYXhHx7Hoo5x68:b-ojug6l) ([raw file](https://github.com/robotlocomotion/drake/blob/1faa309e4a3f83aaba3ea1508a6d68f273fa76a8/drake/multibody/multibody_tree/multibody_tree.h#L400)):*
<details><summary><i>Previously, edrumwri (Evan Drumwright) wrote…</i></summary><blockquote>

BTW Nice!
</blockquote></details>

BTW glad you like it.

---

*[drake/multibody/multibody_tree/multibody_tree.h, line 781 at r5](https://reviewable.io:443/reviews/robotlocomotion/drake/6957#-Kuail_25CbivuPwXyCo:-KuuAlRIBDxhzNzyAdxx:b-z9rpe8) ([raw file](https://github.com/robotlocomotion/drake/blob/1faa309e4a3f83aaba3ea1508a6d68f273fa76a8/drake/multibody/multibody_tree/multibody_tree.h#L781)):*
<details><summary><i>Previously, edrumwri (Evan Drumwright) wrote…</i></summary><blockquote>

Again, why can't they be a function of time?
</blockquote></details>

Yes. They can. I updated doc.

---

*[drake/multibody/multibody_tree/multibody_tree.h, line 1033 at r5](https://reviewable.io:443/reviews/robotlocomotion/drake/6957#-KuajdF85NnNrxB3bln7:-KuuB1_q8I9O2-rXCznF:b-o0ft6i) ([raw file](https://github.com/robotlocomotion/drake/blob/1faa309e4a3f83aaba3ea1508a6d68f273fa76a8/drake/multibody/multibody_tree/multibody_tree.h#L1033)):*
<details><summary><i>Previously, edrumwri (Evan Drumwright) wrote…</i></summary><blockquote>

Any reason you don't return a const reference as you did with AddForceElement() above? Actually, given that this is a helper method, why return anything at all?
</blockquote></details>

Fair enough. I made it `void`

---

*[drake/multibody/multibody_tree/uniform_gravity_field_element.h, line 36 at r5](https://reviewable.io:443/reviews/robotlocomotion/drake/6957#-KuakZnqtO9q4BPrwM8M:-KuuBOYLPVmm5hFsy370:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/1faa309e4a3f83aaba3ea1508a6d68f273fa76a8/drake/multibody/multibody_tree/uniform_gravity_field_element.h#L36)):*
<details><summary><i>Previously, edrumwri (Evan Drumwright) wrote…</i></summary><blockquote>

I'll posit that `g_W` violates the style guide on naming conventions.
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/6957)*
<!-- Sent from Reviewable.io -->

edrumwri commented:
Review status: 8 of 17 files reviewed at latest revision, 9 unresolved discussions.

---

*[drake/multibody/benchmarks/acrobot/acrobot.cc, line 96 at r5](https://reviewable.io:443/reviews/robotlocomotion/drake/6957#-Ku_MN-FLankStt4jesD:-Kuuk2WkE5gQ-KXHQkKO:bpjordl) ([raw file](https://github.com/robotlocomotion/drake/blob/1faa309e4a3f83aaba3ea1508a6d68f273fa76a8/drake/multibody/benchmarks/acrobot/acrobot.cc#L96)):*
<details><summary><i>Previously, amcastro-tri (Alejandro Castro) wrote…</i></summary><blockquote>

People seem to also like lower case `g`, though in texts it usually is a bold g. I think `g(q)` is ok. What do you think?
</blockquote></details>

As noted above (with less specificity), g(q) is used for constraint functions. How about fᵍʳᵃᵛ (which helpfully points out that it's a force), if you must point out gravity specifically? I could live with g(q) if you're not crazy about that, but we may end up with a hard-to-resolve name clash at some point.

---

*[drake/multibody/multibody_tree/body.h, line 172 at r5](https://reviewable.io:443/reviews/robotlocomotion/drake/6957#-Ku_N3ViVavLwdn8fgNt:-KuumbiABVmzYJL0hXtk:bxyt4r6) ([raw file](https://github.com/robotlocomotion/drake/blob/1faa309e4a3f83aaba3ea1508a6d68f273fa76a8/drake/multibody/multibody_tree/body.h#L172)):*
<details><summary><i>Previously, amcastro-tri (Alejandro Castro) wrote…</i></summary><blockquote>

It might be a parameter of your system say, for sys ID. So far only the API is implemented. 
</blockquote></details>

Wouldn't that go into your context though?

---

*[drake/multibody/multibody_tree/force_element.h, line 140 at r5](https://reviewable.io:443/reviews/robotlocomotion/drake/6957#-KuahkkI64hV7ZqLnPzo:-Kuumxi9RBFmbMLYKQaX:b6fn84u) ([raw file](https://github.com/robotlocomotion/drake/blob/1faa309e4a3f83aaba3ea1508a6d68f273fa76a8/drake/multibody/multibody_tree/force_element.h#L140)):*
<details><summary><i>Previously, amcastro-tri (Alejandro Castro) wrote…</i></summary><blockquote>

Actually these should be named `DoCalcXXX()` and be private. The `CalcXXX()` would only thake the `context` as argument. However we still do not have caching. What do you suggest?
</blockquote></details>

Am I correct in assuming that after Sherm gets the caching in, this issue will resolve itself by allowing you to turn this method into DoCalcConservativePower()?

---

*[drake/multibody/multibody_tree/force_element.h, line 185 at r5](https://reviewable.io:443/reviews/robotlocomotion/drake/6957#-KuahRnklbXAWc-0uSKv:-KuuwHrX5xOPjZXYKtNZ:b-322ugb) ([raw file](https://github.com/robotlocomotion/drake/blob/1faa309e4a3f83aaba3ea1508a6d68f273fa76a8/drake/multibody/multibody_tree/force_element.h#L185)):*
<details><summary><i>Previously, amcastro-tri (Alejandro Castro) wrote…</i></summary><blockquote>

I'd prefer not to do so. The `DoCloneToScalar() mechanism I have right now will hopefully get replaced by something similar to what we have in systems, which will also hopefully allow us to template on novel scalar types without having to change update every  source file. Therefore, I'd prefer not leaking this into the main class's API.
</blockquote></details>

Regardless, you're requiring the implementer to implement these and giving very little information on doing so. You need some documentation that describes what the implementer need do.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/6957)*
<!-- Sent from Reviewable.io -->

sherm1 commented:
Review status: 8 of 17 files reviewed at latest revision, 9 unresolved discussions.

---

*[drake/multibody/benchmarks/acrobot/acrobot.cc, line 96 at r5](https://reviewable.io:443/reviews/robotlocomotion/drake/6957#-Ku_MN-FLankStt4jesD:-KuuzgEATU1rpeDd39OW:b-fgmsa4) ([raw file](https://github.com/robotlocomotion/drake/blob/1faa309e4a3f83aaba3ea1508a6d68f273fa76a8/drake/multibody/benchmarks/acrobot/acrobot.cc#L96)):*
<details><summary><i>Previously, edrumwri (Evan Drumwright) wrote…</i></summary><blockquote>

As noted above (with less specificity), g(q) is used for constraint functions. How about fᵍʳᵃᵛ (which helpfully points out that it's a force), if you must point out gravity specifically? I could live with g(q) if you're not crazy about that, but we may end up with a hard-to-resolve name clash at some point.
</blockquote></details>

:+1: for using "f" to mean "force"!

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/6957)*
<!-- Sent from Reviewable.io -->

amcastro-tri commented:
Added the requested unit tests @edrumwri. PTAL.

---

Review status: 6 of 17 files reviewed at latest revision, 6 unresolved discussions.

---

*[drake/multibody/benchmarks/acrobot/acrobot.cc, line 96 at r5](https://reviewable.io:443/reviews/robotlocomotion/drake/6957#-Ku_MN-FLankStt4jesD:-Kuv6ce5eLKljmb2FS43:b-wwqq5q) ([raw file](https://github.com/robotlocomotion/drake/blob/1faa309e4a3f83aaba3ea1508a6d68f273fa76a8/drake/multibody/benchmarks/acrobot/acrobot.cc#L96)):*
<details><summary><i>Previously, sherm1 (Michael Sherman) wrote…</i></summary><blockquote>

:+1: for using "f" to mean "force"!
</blockquote></details>

we use `f` for linear forces. What about `tau_g(q)`?

---

*[drake/multibody/multibody_tree/force_element.h, line 140 at r5](https://reviewable.io:443/reviews/robotlocomotion/drake/6957#-KuahkkI64hV7ZqLnPzo:-Kuv61xMD-ayDNnpGkq0:btqqu5l) ([raw file](https://github.com/robotlocomotion/drake/blob/1faa309e4a3f83aaba3ea1508a6d68f273fa76a8/drake/multibody/multibody_tree/force_element.h#L140)):*
<details><summary><i>Previously, edrumwri (Evan Drumwright) wrote…</i></summary><blockquote>

Am I correct in assuming that after Sherm gets the caching in, this issue will resolve itself by allowing you to turn this method into DoCalcConservativePower()?
</blockquote></details>

Correct.

---

*[drake/multibody/multibody_tree/force_element.h, line 185 at r5](https://reviewable.io:443/reviews/robotlocomotion/drake/6957#-KuahRnklbXAWc-0uSKv:-Kuv6MWYVIrwjrkaoHJo:bx2231o) ([raw file](https://github.com/robotlocomotion/drake/blob/1faa309e4a3f83aaba3ea1508a6d68f273fa76a8/drake/multibody/multibody_tree/force_element.h#L185)):*
<details><summary><i>Previously, edrumwri (Evan Drumwright) wrote…</i></summary><blockquote>

Regardless, you're requiring the implementer to implement these and giving very little information on doing so. You need some documentation that describes what the implementer need do.
</blockquote></details>

Maybe in an upcoming PR. Not sure at this point if worth spending time on documenting the soon-to-be-deprecated API (more than API, internals actually)

---

*[drake/multibody/multibody_tree/uniform_gravity_field_element.cc, line 103 at r5](https://reviewable.io:443/reviews/robotlocomotion/drake/6957#-KuawCBjKXEzSF-gtaLJ:-Kuv6_fTAD15F6Gcs70f:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/1faa309e4a3f83aaba3ea1508a6d68f273fa76a8/drake/multibody/multibody_tree/uniform_gravity_field_element.cc#L103)):*
<details><summary><i>Previously, edrumwri (Evan Drumwright) wrote…</i></summary><blockquote>

Please add a test of this function.
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/6957#-:-Kuv6tDfp14yJN_Dzb4e:bz9s96n)*
<!-- Sent from Reviewable.io -->

amcastro-tri commented:
Review status: 6 of 17 files reviewed at latest revision, 6 unresolved discussions.

---

*[drake/multibody/benchmarks/acrobot/acrobot.cc, line 96 at r5](https://reviewable.io:443/reviews/robotlocomotion/drake/6957#-Ku_MN-FLankStt4jesD:-KuyCVIQv214HBzsJN9_:blwce5k) ([raw file](https://github.com/robotlocomotion/drake/blob/1faa309e4a3f83aaba3ea1508a6d68f273fa76a8/drake/multibody/benchmarks/acrobot/acrobot.cc#L96)):*
<details><summary><i>Previously, amcastro-tri (Alejandro Castro) wrote…</i></summary><blockquote>

we use `f` for linear forces. What about `tau_g(q)`?
</blockquote></details>

Renamed to `tau_g(q)` everywhere. 

---

*[drake/multibody/benchmarks/acrobot/acrobot.h, line 80 at r5](https://reviewable.io:443/reviews/robotlocomotion/drake/6957#-Ku_MsmBxxDowamOcVCo:-KuyC_ScinhYgWLp5U5S:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/1faa309e4a3f83aaba3ea1508a6d68f273fa76a8/drake/multibody/benchmarks/acrobot/acrobot.h#L80)):*
<details><summary><i>Previously, edrumwri (Evan Drumwright) wrote…</i></summary><blockquote>

Same name collision with G(q) here.
</blockquote></details>

Done.

---

*[drake/multibody/multibody_tree/body.h, line 172 at r5](https://reviewable.io:443/reviews/robotlocomotion/drake/6957#-Ku_N3ViVavLwdn8fgNt:-KuyAly02yq1MIcFVDqA:b-f152iq) ([raw file](https://github.com/robotlocomotion/drake/blob/1faa309e4a3f83aaba3ea1508a6d68f273fa76a8/drake/multibody/multibody_tree/body.h#L172)):*
<details><summary><i>Previously, edrumwri (Evan Drumwright) wrote…</i></summary><blockquote>

Wouldn't that go into your context though?
</blockquote></details>

That's actually a good point. What have we been calling the methods used to retrieve parameters? do we have examples already? are those something like `T get_mass(context)`?
I think @sherm1 at some point liked the idea of calling these `double get_default_mass()` and `T get_mass(context)`, is that what we'd like?

---

*[drake/multibody/multibody_tree/force_element.h, line 185 at r5](https://reviewable.io:443/reviews/robotlocomotion/drake/6957#-KuahRnklbXAWc-0uSKv:-KuyKDp3CggU6UninelW:b-p431qh) ([raw file](https://github.com/robotlocomotion/drake/blob/1faa309e4a3f83aaba3ea1508a6d68f273fa76a8/drake/multibody/multibody_tree/force_element.h#L185)):*
<details><summary><i>Previously, amcastro-tri (Alejandro Castro) wrote…</i></summary><blockquote>

Maybe in an upcoming PR. Not sure at this point if worth spending time on documenting the soon-to-be-deprecated API (more than API, internals actually)
</blockquote></details>

Never mind. That future PR is not in the near future so probably much better I document this here. Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/6957)*
<!-- Sent from Reviewable.io -->

edrumwri commented:
Why the replacement from Isometry3d::rotation() to Isometry3D::linear()

---

Reviewed 3 of 11 files at r6.
Review status: 9 of 17 files reviewed at latest revision, 9 unresolved discussions.

---

*[drake/multibody/benchmarks/acrobot/acrobot.cc, line 96 at r5](https://reviewable.io:443/reviews/robotlocomotion/drake/6957#-Ku_MN-FLankStt4jesD:-KuyYKkI5SxId_cGUevy:b39truv) ([raw file](https://github.com/robotlocomotion/drake/blob/1faa309e4a3f83aaba3ea1508a6d68f273fa76a8/drake/multibody/benchmarks/acrobot/acrobot.cc#L96)):*
<details><summary><i>Previously, amcastro-tri (Alejandro Castro) wrote…</i></summary><blockquote>

Renamed to `tau_g(q)` everywhere. 
</blockquote></details>

I'll take it.

---

*[drake/multibody/multibody_tree/force_element.h, line 140 at r5](https://reviewable.io:443/reviews/robotlocomotion/drake/6957#-KuahkkI64hV7ZqLnPzo:-KuyYZxoiztqSrRL7LQd:b-jquixs) ([raw file](https://github.com/robotlocomotion/drake/blob/1faa309e4a3f83aaba3ea1508a6d68f273fa76a8/drake/multibody/multibody_tree/force_element.h#L140)):*
<details><summary><i>Previously, amcastro-tri (Alejandro Castro) wrote…</i></summary><blockquote>

Correct.
</blockquote></details>

Do you want to point out that you'll address this in a TODO so that you don't have to wait on that?

---

*[drake/multibody/multibody_tree/force_element.h, line 171 at r6](https://reviewable.io:443/reviews/robotlocomotion/drake/6957#-KuyZNNqevwOAJVcKVlT:-KuyZNNqevwOAJVcKVlU:b-fd3nr2) ([raw file](https://github.com/robotlocomotion/drake/blob/d909789f45ac71367d7b802db816537ed39219ed/drake/multibody/multibody_tree/force_element.h#L171)):*
> ```objc
> 
>  protected:
>   /// Implementation of NVI CalcAndAddForceContribution().
> ```

This isn't an implementation but an interface, and you still need to tell the implementer exactly what it is they are supposed to do. You can point to CalcAndAddForceContribution() for help.

---

*[drake/multibody/multibody_tree/force_element.h, line 189 at r6](https://reviewable.io:443/reviews/robotlocomotion/drake/6957#-KuyZl4ue2CxkjBjHBUo:-KuyZl4yPsqaUCFWq2JE:bfm7ux1) ([raw file](https://github.com/robotlocomotion/drake/blob/d909789f45ac71367d7b802db816537ed39219ed/drake/multibody/multibody_tree/force_element.h#L189)):*
> ```objc
>   /// Specific force element subclasses must implement these to support scalar
>   /// conversion to other types. These methods are only called from
>   /// MultibodyTree::CloneToScalar(), users _must_ not call these explicitely.
> ```

, -> ;

---

*[drake/multibody/multibody_tree/force_element.h, line 189 at r6](https://reviewable.io:443/reviews/robotlocomotion/drake/6957#-KuyZpsVDGEbHmscLwen:-KuyZpsXo1w5ghZkb03S:b-176vd0) ([raw file](https://github.com/robotlocomotion/drake/blob/d909789f45ac71367d7b802db816537ed39219ed/drake/multibody/multibody_tree/force_element.h#L189)):*
> ```objc
>   /// Specific force element subclasses must implement these to support scalar
>   /// conversion to other types. These methods are only called from
>   /// MultibodyTree::CloneToScalar(), users _must_ not call these explicitely.
> ```

explictely -> explicitly

---

*[drake/multibody/multibody_tree/force_element.h, line 189 at r6](https://reviewable.io:443/reviews/robotlocomotion/drake/6957#-KuyZtPV4dSx6oAEAbTG:-KuyZtPXCoap5yMW2sOe:b-lxy42d) ([raw file](https://github.com/robotlocomotion/drake/blob/d909789f45ac71367d7b802db816537ed39219ed/drake/multibody/multibody_tree/force_element.h#L189)):*
> ```objc
>   /// Specific force element subclasses must implement these to support scalar
>   /// conversion to other types. These methods are only called from
>   /// MultibodyTree::CloneToScalar(), users _must_ not call these explicitely.
> ```

"must not" or "should not need to"? Must not implies something bad is going to happen if they do.

---

*[drake/multibody/multibody_tree/force_element.h, line 197 at r6](https://reviewable.io:443/reviews/robotlocomotion/drake/6957#-Kuy_9OrF2dPLzzFM9zZ:-Kuy_9OrF2dPLzzFM9z_:buuv7hl) ([raw file](https://github.com/robotlocomotion/drake/blob/d909789f45ac71367d7b802db816537ed39219ed/drake/multibody/multibody_tree/force_element.h#L197)):*
> ```objc
>   /// ForceElement::DoCloneToScalar() can retrieve clones from `tree_clone` as
>   /// needed.
>   /// Consider the following example for a `SpringElement<T>`:
> ```

The code block is much appreciated. However, as you introduce it can you briefly tell the reader what the example will illustrate so that they don't have to read through the block of code to find out?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/6957#-:-KuynrS4THyV7hZ55pZR:b2sw0x0)*
<!-- Sent from Reviewable.io -->

amcastro-tri commented:
Because I almost re-introduce the bug I tried to fix in #6989.
Also, in this case using `.rotation()` broke my `AutoDiffXd` unit tests!!  that is because there is a lot of math going with `.rotation()` to extract a valid rotation matrix and it seems like it fails for very particular values when using `AutoDiffXd`. I could identify exactly why, but clearly using SVD to get a rotation matrix from an already valid rotation matrix is a killer!

---

Review status: 9 of 17 files reviewed at latest revision, 7 unresolved discussions.

---

*[drake/multibody/multibody_tree/force_element.h, line 140 at r5](https://reviewable.io:443/reviews/robotlocomotion/drake/6957#-KuahkkI64hV7ZqLnPzo:-Kuyr22GU4Ygz6Df7GKp:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/1faa309e4a3f83aaba3ea1508a6d68f273fa76a8/drake/multibody/multibody_tree/force_element.h#L140)):*
<details><summary><i>Previously, edrumwri (Evan Drumwright) wrote…</i></summary><blockquote>

Do you want to point out that you'll address this in a TODO so that you don't have to wait on that?
</blockquote></details>

Done.

---

*[drake/multibody/multibody_tree/force_element.h, line 171 at r6](https://reviewable.io:443/reviews/robotlocomotion/drake/6957#-KuyZNNqevwOAJVcKVlT:-Kuys--vm0f7JDHeKf9B:bb7oryy) ([raw file](https://github.com/robotlocomotion/drake/blob/d909789f45ac71367d7b802db816537ed39219ed/drake/multibody/multibody_tree/force_element.h#L171)):*
<details><summary><i>Previously, edrumwri (Evan Drumwright) wrote…</i></summary><blockquote>

This isn't an implementation but an interface, and you still need to tell the implementer exactly what it is they are supposed to do. You can point to CalcAndAddForceContribution() for help.
</blockquote></details>

sure

---

*[drake/multibody/multibody_tree/force_element.h, line 189 at r6](https://reviewable.io:443/reviews/robotlocomotion/drake/6957#-KuyZl4ue2CxkjBjHBUo:-KuysCbK50fXkA2-9MFe:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/d909789f45ac71367d7b802db816537ed39219ed/drake/multibody/multibody_tree/force_element.h#L189)):*
<details><summary><i>Previously, edrumwri (Evan Drumwright) wrote…</i></summary><blockquote>

, -> ;
</blockquote></details>

Done.

---

*[drake/multibody/multibody_tree/force_element.h, line 189 at r6](https://reviewable.io:443/reviews/robotlocomotion/drake/6957#-KuyZpsVDGEbHmscLwen:-KuysHoXY1Yr9Ktl_JWJ:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/d909789f45ac71367d7b802db816537ed39219ed/drake/multibody/multibody_tree/force_element.h#L189)):*
<details><summary><i>Previously, edrumwri (Evan Drumwright) wrote…</i></summary><blockquote>

explictely -> explicitly
</blockquote></details>

Done.

---

*[drake/multibody/multibody_tree/force_element.h, line 189 at r6](https://reviewable.io:443/reviews/robotlocomotion/drake/6957#-KuyZtPV4dSx6oAEAbTG:-KuysLSBmzMZFMhmkaHq:b-1wur6l) ([raw file](https://github.com/robotlocomotion/drake/blob/d909789f45ac71367d7b802db816537ed39219ed/drake/multibody/multibody_tree/force_element.h#L189)):*
<details><summary><i>Previously, edrumwri (Evan Drumwright) wrote…</i></summary><blockquote>

"must not" or "should not need to"? Must not implies something bad is going to happen if they do.
</blockquote></details>

"must not". There is no way for them to guarantee they have a `tree_clone` in a valid state, while `MultibodyTree::CloneToScalar()` does exactly that for them.

---

*[drake/multibody/multibody_tree/force_element.h, line 197 at r6](https://reviewable.io:443/reviews/robotlocomotion/drake/6957#-Kuy_9OrF2dPLzzFM9zZ:-KuysnOW2kK5_jIae0yz:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/d909789f45ac71367d7b802db816537ed39219ed/drake/multibody/multibody_tree/force_element.h#L197)):*
<details><summary><i>Previously, edrumwri (Evan Drumwright) wrote…</i></summary><blockquote>

The code block is much appreciated. However, as you introduce it can you briefly tell the reader what the example will illustrate so that they don't have to read through the block of code to find out?
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/6957#-:-Kuypw_vXjKqidcaT6ba:bqbt8b9)*
<!-- Sent from Reviewable.io -->

amcastro-tri commented:
+@sammy-tri for platform review please per regular rotation.

---

Review status: 8 of 17 files reviewed at latest revision, 7 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/6957#-:-KuytKkU6-R-R1RDBsIV:bm4vvti)*
<!-- Sent from Reviewable.io -->

sherm1 commented:
Review status: 8 of 17 files reviewed at latest revision, 7 unresolved discussions.

---

*[drake/multibody/multibody_tree/body.h, line 172 at r5](https://reviewable.io:443/reviews/robotlocomotion/drake/6957#-Ku_N3ViVavLwdn8fgNt:-KuytCE49Ew8l1UAnCIf:b-t4jbvv) ([raw file](https://github.com/robotlocomotion/drake/blob/1faa309e4a3f83aaba3ea1508a6d68f273fa76a8/drake/multibody/multibody_tree/body.h#L172)):*
<details><summary><i>Previously, amcastro-tri (Alejandro Castro) wrote…</i></summary><blockquote>

That's actually a good point. What have we been calling the methods used to retrieve parameters? do we have examples already? are those something like `T get_mass(context)`?
I think @sherm1 at some point liked the idea of calling these `double get_default_mass()` and `T get_mass(context)`, is that what we'd like?
</blockquote></details>

BTW I still like `double get_default_mass()` and `T get_mass(context)`.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/6957)*
<!-- Sent from Reviewable.io -->

sammy-tri commented:
First pass complete.

---

Reviewed 1 of 26 files at r1, 1 of 12 files at r2, 2 of 4 files at r3, 1 of 13 files at r4, 1 of 5 files at r5, 10 of 11 files at r6, 1 of 1 files at r7.
Review status: all files reviewed at latest revision, 9 unresolved discussions.

---

*[drake/multibody/multibody_tree/multibody_tree.h, line 390 at r7](https://reviewable.io:443/reviews/robotlocomotion/drake/6957#-KuzKI4aa9qEhHDRBBbS:-KuzKI4bF50J3U-GmJiy:b-ixxtt5) ([raw file](https://github.com/robotlocomotion/drake/blob/8de0bd2370574cd708c7760b4b664fbfd706112a/drake/multibody/multibody_tree/multibody_tree.h#L390)):*
> ```objc
>     if (topology_is_valid()) {
>       throw std::logic_error(
>           "This MultiforceTree is finalized already. Therefore adding more "
> ```

The other `AddFoo` functions throw with a message starting with "This MuiltibodyTree is finalized already".  Is the spelling "MultiforceTree" deliberate here?

---

*[drake/multibody/multibody_tree/multibody_tree.h, line 1072 at r7](https://reviewable.io:443/reviews/robotlocomotion/drake/6957#-KuzKwyhb-XPXV5fFs6O:-KuzKwyiJl2hpDRoo3HC:b-zgivz4) ([raw file](https://github.com/robotlocomotion/drake/blob/8de0bd2370574cd708c7760b4b664fbfd706112a/drake/multibody/multibody_tree/multibody_tree.h#L1072)):*
> ```objc
>       EigenPtr<VectorX<T>> Cv) const;
> 
>   // Implementation of DoCalcPotentialEnergy().
> ```

Should this say "Implementation of CalcPotentialEnergy"?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/6957#-:-KuzM4AAV7IML9ZxSwV7:bvb1a6z)*
<!-- Sent from Reviewable.io -->

edrumwri commented:
Multiple minor open comments to address and then <img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> :argentina: 

---

Reviewed 7 of 11 files at r6, 1 of 1 files at r7.
Review status: all files reviewed at latest revision, 9 unresolved discussions.

---

*[drake/multibody/multibody_tree/force_element.h, line 171 at r6](https://reviewable.io:443/reviews/robotlocomotion/drake/6957#-KuyZNNqevwOAJVcKVlT:-KuzQJYrWY6qt1A34CKI:b6sjhpf) ([raw file](https://github.com/robotlocomotion/drake/blob/d909789f45ac71367d7b802db816537ed39219ed/drake/multibody/multibody_tree/force_element.h#L171)):*
<details><summary><i>Previously, amcastro-tri (Alejandro Castro) wrote…</i></summary><blockquote>

sure
</blockquote></details>

BTW I suggest changing "for details" to "for details describing what this function should compute and the function's parameters". (Otherwise, it's unclear which details you are pointing the implementer to).

---

*[drake/multibody/multibody_tree/force_element.h, line 189 at r6](https://reviewable.io:443/reviews/robotlocomotion/drake/6957#-KuyZtPV4dSx6oAEAbTG:-KuzQctV1T0fryfv2L_b:b144hu) ([raw file](https://github.com/robotlocomotion/drake/blob/d909789f45ac71367d7b802db816537ed39219ed/drake/multibody/multibody_tree/force_element.h#L189)):*
<details><summary><i>Previously, amcastro-tri (Alejandro Castro) wrote…</i></summary><blockquote>

"must not". There is no way for them to guarantee they have a `tree_clone` in a valid state, while `MultibodyTree::CloneToScalar()` does exactly that for them.
</blockquote></details>

Ok, please explain that explicitly (I know you go into details of what the cloning process does below, but I didn't understand that they shouldn't call it for the particular reason that the structure wouldn't be recreated properly.)

---

*[drake/multibody/multibody_tree/force_element.h, line 197 at r6](https://reviewable.io:443/reviews/robotlocomotion/drake/6957#-Kuy_9OrF2dPLzzFM9zZ:-KuzQtcu0gxGAm-dADeT:b90h4pu) ([raw file](https://github.com/robotlocomotion/drake/blob/d909789f45ac71367d7b802db816537ed39219ed/drake/multibody/multibody_tree/force_element.h#L197)):*
<details><summary><i>Previously, amcastro-tri (Alejandro Castro) wrote…</i></summary><blockquote>

Done.
</blockquote></details>

That wasn't what I meant- I meant, what do you hope for the reader of this code to get out of this example? Just an example of how DoCloneToScalar() might be implemented?

---

*[drake/multibody/multibody_tree/multibody_tree.h, line 841 at r7](https://reviewable.io:443/reviews/robotlocomotion/drake/6957#-KuzRUJng-un9dpvfdK5:-KuzRUJng-un9dpvfdK6:bfhyirz) ([raw file](https://github.com/robotlocomotion/drake/blob/8de0bd2370574cd708c7760b4b664fbfd706112a/drake/multibody/multibody_tree/multibody_tree.h#L841)):*
> ```objc
>   /// Computes and returns the power generated by conservative forces in the
>   /// multibody model. This quantity is defined to be positive when the
>   /// potential energy is decreasing. In other words, if `PE` is the potential
> ```

BTW, if you rename `PE` to `U`, you can avoid the d/dt notation and parentheses and instead use the far more readable U̇.

---

*[drake/multibody/multibody_tree/multibody_tree.h, line 1073 at r7](https://reviewable.io:443/reviews/robotlocomotion/drake/6957#-KuzRm0mHkJFelN4fhA9:-KuzRm0mHkJFelN4fhAA:blc3iij) ([raw file](https://github.com/robotlocomotion/drake/blob/8de0bd2370574cd708c7760b4b664fbfd706112a/drake/multibody/multibody_tree/multibody_tree.h#L1073)):*
> ```objc
> 
>   // Implementation of DoCalcPotentialEnergy().
>   // It is assumed that the position kinematics cache pc is in sync with
> ```

BTW nit "pc" -> "`pc" (here and below)

---

*[drake/multibody/multibody_tree/test/double_pendulum_test.cc, line 1181 at r7](https://reviewable.io:443/reviews/robotlocomotion/drake/6957#-KuzS882yxDXSBpRtgF3:-KuzS882yxDXSBpRtgF4:bmcuzic) ([raw file](https://github.com/robotlocomotion/drake/blob/8de0bd2370574cd708c7760b4b664fbfd706112a/drake/multibody/multibody_tree/test/double_pendulum_test.cc#L1181)):*
> ```cpp
>           EXPECT_TRUE(V_WLcm.IsApprox(V_WLcm_expected, kTolerance));
> 
>           // Compute potential energy, and its time derivative.
> ```

BTW Nice!

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/6957#-:-KuzSOVHd3wG2Z9-xGWD:b-2ltdda)*
<!-- Sent from Reviewable.io -->

amcastro-tri commented:
Review status: all files reviewed at latest revision, 9 unresolved discussions.

---

*[drake/multibody/multibody_tree/body.h, line 172 at r5](https://reviewable.io:443/reviews/robotlocomotion/drake/6957#-Ku_N3ViVavLwdn8fgNt:-KuzgeDlWaiPGhS9bXpp:b-l4r32a) ([raw file](https://github.com/robotlocomotion/drake/blob/1faa309e4a3f83aaba3ea1508a6d68f273fa76a8/drake/multibody/multibody_tree/body.h#L172)):*
<details><summary><i>Previously, sherm1 (Michael Sherman) wrote…</i></summary><blockquote>

BTW I still like `double get_default_mass()` and `T get_mass(context)`.
</blockquote></details>

Refactored to `get_mass(context)`.

---

*[drake/multibody/multibody_tree/force_element.h, line 171 at r6](https://reviewable.io:443/reviews/robotlocomotion/drake/6957#-KuyZNNqevwOAJVcKVlT:-Kuzh56ISr08NujTE2tJ:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/d909789f45ac71367d7b802db816537ed39219ed/drake/multibody/multibody_tree/force_element.h#L171)):*
<details><summary><i>Previously, edrumwri (Evan Drumwright) wrote…</i></summary><blockquote>

BTW I suggest changing "for details" to "for details describing what this function should compute and the function's parameters". (Otherwise, it's unclear which details you are pointing the implementer to).
</blockquote></details>

Done.

---

*[drake/multibody/multibody_tree/force_element.h, line 189 at r6](https://reviewable.io:443/reviews/robotlocomotion/drake/6957#-KuyZtPV4dSx6oAEAbTG:-KuzhsmCuwXkkvfjGvpN:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/d909789f45ac71367d7b802db816537ed39219ed/drake/multibody/multibody_tree/force_element.h#L189)):*
<details><summary><i>Previously, edrumwri (Evan Drumwright) wrote…</i></summary><blockquote>

Ok, please explain that explicitly (I know you go into details of what the cloning process does below, but I didn't understand that they shouldn't call it for the particular reason that the structure wouldn't be recreated properly.)
</blockquote></details>

Done.

---

*[drake/multibody/multibody_tree/force_element.h, line 197 at r6](https://reviewable.io:443/reviews/robotlocomotion/drake/6957#-Kuy_9OrF2dPLzzFM9zZ:-KuzkFdj4poKEItSiJMa:bjdknjy) ([raw file](https://github.com/robotlocomotion/drake/blob/d909789f45ac71367d7b802db816537ed39219ed/drake/multibody/multibody_tree/force_element.h#L197)):*
<details><summary><i>Previously, edrumwri (Evan Drumwright) wrote…</i></summary><blockquote>

That wasn't what I meant- I meant, what do you hope for the reader of this code to get out of this example? Just an example of how DoCloneToScalar() might be implemented?
</blockquote></details>

Added reference to example in `multibody_tree.h`

---

*[drake/multibody/multibody_tree/multibody_tree.h, line 390 at r7](https://reviewable.io:443/reviews/robotlocomotion/drake/6957#-KuzKI4aa9qEhHDRBBbS:-KuzkVJzgDbOqBjYDdvY:bg9rvsl) ([raw file](https://github.com/robotlocomotion/drake/blob/8de0bd2370574cd708c7760b4b664fbfd706112a/drake/multibody/multibody_tree/multibody_tree.h#L390)):*
<details><summary><i>Previously, sammy-tri (Sam Creasey) wrote…</i></summary><blockquote>

The other `AddFoo` functions throw with a message starting with "This MuiltibodyTree is finalized already".  Is the spelling "MultiforceTree" deliberate here?
</blockquote></details>

Ah! body/force find/replace! good catch!

---

*[drake/multibody/multibody_tree/multibody_tree.h, line 841 at r7](https://reviewable.io:443/reviews/robotlocomotion/drake/6957#-KuzRUJng-un9dpvfdK5:-Kuzl0gRka32ZdMa6-10:b2ywptm) ([raw file](https://github.com/robotlocomotion/drake/blob/8de0bd2370574cd708c7760b4b664fbfd706112a/drake/multibody/multibody_tree/multibody_tree.h#L841)):*
<details><summary><i>Previously, edrumwri (Evan Drumwright) wrote…</i></summary><blockquote>

BTW, if you rename `PE` to `U`, you can avoid the d/dt notation and parentheses and instead use the far more readable U̇.
</blockquote></details>

BTW I like it, thanks

---

*[drake/multibody/multibody_tree/multibody_tree.h, line 1072 at r7](https://reviewable.io:443/reviews/robotlocomotion/drake/6957#-KuzKwyhb-XPXV5fFs6O:-KuzlD2NeyOR9Zjy_4ka:b-mvv2sx) ([raw file](https://github.com/robotlocomotion/drake/blob/8de0bd2370574cd708c7760b4b664fbfd706112a/drake/multibody/multibody_tree/multibody_tree.h#L1072)):*
<details><summary><i>Previously, sammy-tri (Sam Creasey) wrote…</i></summary><blockquote>

Should this say "Implementation of CalcPotentialEnergy"?
</blockquote></details>

yes, thanks.

---

*[drake/multibody/multibody_tree/multibody_tree.h, line 1073 at r7](https://reviewable.io:443/reviews/robotlocomotion/drake/6957#-KuzRm0mHkJFelN4fhA9:-KuzlMH_gnCN4zvghzeD:b-3m4pul) ([raw file](https://github.com/robotlocomotion/drake/blob/8de0bd2370574cd708c7760b4b664fbfd706112a/drake/multibody/multibody_tree/multibody_tree.h#L1073)):*
<details><summary><i>Previously, edrumwri (Evan Drumwright) wrote…</i></summary><blockquote>

BTW nit "pc" -> "`pc" (here and below)
</blockquote></details>

It won't make a difference since these are not Doxygen comments.

---

*[drake/multibody/multibody_tree/test/double_pendulum_test.cc, line 1181 at r7](https://reviewable.io:443/reviews/robotlocomotion/drake/6957#-KuzS882yxDXSBpRtgF3:-KuzlWuwi-JrBGTK0i1W:bah7epd) ([raw file](https://github.com/robotlocomotion/drake/blob/8de0bd2370574cd708c7760b4b664fbfd706112a/drake/multibody/multibody_tree/test/double_pendulum_test.cc#L1181)):*
<details><summary><i>Previously, edrumwri (Evan Drumwright) wrote…</i></summary><blockquote>

BTW Nice!
</blockquote></details>

right? being able to pass autodiff is just awesome!

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/6957)*
<!-- Sent from Reviewable.io -->

edrumwri commented:
Reviewed 2 of 5 files at r8.
Review status: 14 of 17 files reviewed at latest revision, 3 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/6957)*
<!-- Sent from Reviewable.io -->

sammy-tri commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Reviewed 5 of 5 files at r8.
Review status: all files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/6957#-:-Kuzx_JiT7WIvIc-iRzg:bnfp4nl)*
<!-- Sent from Reviewable.io -->
